efi/preinstall/checks_context.go

```diff
@@ -0,0 +1,585 @@
+			ActionRebootToFWSettings, // suggest rebooting to the firmware settings UI to properly configure secure boot
+			// TODO: Add action to add PCRProfileOptionPermitNoSecureBootPolicyProfile to the PCRProfileOptionsFlags,
+			//  noting that this flag will need to be subsequently passed to WithAutoTCGPCRProfile.
+		},
+		ErrorKindWeakSecureBootAlgorithmsDetected: []Action{
+			// TODO: Add action to add PermitWeakSecureBootAlgorithms to CheckFlags.
+		},
+		ErrorKindPreOSDigestVerificationDetected: []Action{
+			// TODO: Add action to add PermitPreOSVerificationUsingDigests to CheckFlags.
+		},
+	}
+}
+
+// RunChecksContext maintains context for multiple invocations of [RunChecks] to permit the
+// install process to iterate and resolve detected issues where possible. It also reduces
+// the burden of selecting an initial set of [CheckFlags].
+type RunChecksContext struct {
+	env          internal_efi.HostEnvironment
+	flags        CheckFlags
+	loadedImages []secboot_efi.Image
+	profileOpts  PCRProfileOptionsFlags
+
+	errs    []error
+	lastErr error
+	result  *CheckResult
+
+	availableActions map[Action]bool
+	expectedActions  []Action
+}
+
+// NewRunChecksContext returns a new RunChecksContext instance with the initial flags for [RunChecks]
+// and the supplied list of boot components for the current boot (see the documentation for [RunChecks].
+// The supplied [PCRProfileOptionsFlags] represent the preferred flags for [WithAutoTCGPCRProfile] and
+// should match the flags later passed to this function when creating a PCR profile. The PCRs that are
+// determined to be required to build the profile using [WithAutoTCGPCRProfile] will be made mandatory
+// automatically by this function by passing the relevant flags to [RunChecks].
+//
+// As [PCRProfileOptionsFlags] is intended to provide some limited amount of user customisation, this
+// API [RunChecksContext] should be executed again with the new set of flags, should the user wish to
+// change them. In this case, the caller should pass the PostInstallChecks flag as an initial flag.
+//
+// There is no need for the caller to supply any of these *SupportRequired flags as the initial flags,
+// and this may have the effect of limiting the number of devices which pass the checks.
+func NewRunChecksContext(initialFlags CheckFlags, loadedImages []secboot_efi.Image, profileOpts PCRProfileOptionsFlags) *RunChecksContext {
+	return &RunChecksContext{
+		env:          runChecksEnv,
+		flags:        initialFlags,
+		loadedImages: loadedImages,
+		profileOpts:  profileOpts,
+		// Populate actions that are always available or available by default
+		// unless we discover that they aren't later on.
+		availableActions: map[Action]bool{
+			ActionNone:               true,
+			ActionReboot:             true,
+			ActionShutdown:           true,
+			ActionRebootToFWSettings: true,
+			ActionContactOEM:         true,
+			ActionContactOSVendor:    true,
+		},
+		expectedActions: []Action{ActionNone},
```

Comment:
do we need to set this as is also set in Run?

Reply:
It's reset in `Run` after running the supplied action - it's set initially here to bootstrap it otherwise `Run` would fail with an unexpected action. It might be better to just clear `expectedActions` and have `ActionNone` as implicitly accepted instead.

Reply:
I've changed the semantics of `expectedActions` now so that `ActionNone` is always expected, even when set to nil.